src/inc/til/bitmap.h

```diff
@@ -246,22 +397,79 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
 
         constexpr bool none() const noexcept
         {
-            return _dirty.empty();
+            return !_dirty.has_value();
         }
 
         constexpr bool all() const noexcept
         {
-            return _dirty == _rc;
+            return _dirty.has_value() && _dirty.value() == _rc;
+        }
+
+        std::wstring to_string() const
+        {
+            std::wstringstream wss;
+            wss << std::endl
+                << L"Bitmap of size " << _sz.to_string() << " contains the following dirty regions:" << std::endl;
+            wss << L"Runs:" << std::endl;
+
+            for (auto& item : *this)
+            {
+                wss << L"\t- " << item.to_string() << std::endl;
+            }
+
+            return wss.str();
         }
 
     private:
-        til::rectangle _dirty;
+        std::optional<til::rectangle> _dirty;

```

Comment:
why'd dirty become an optional? a sizeless rect is not a rect, so it has an optional state. this introduces the chance that the optional is populated but with 0,0,0,0

Reply:
Oh. Well, I did it because I wanted to make sure that the first invalidation added would overwrite the `_dirty` where as all subsequent ones would union `|=` to it. 

Prior to... yesterday... I was doing union `|=` no matter what, which means that every `_dirty` was including the top-left corner as that was the empty/default state.

I suppose I could either...
1. Make it check `_dirty.empty()` and then do `=` if true to overwrite it and |= if false.
2. I could make `|=` be an assignment if the rectangle was already empty.

**NOTE** As I wrote this, I realize that #2 is what it already was doing. And this whole thing was unnecessary. So I'm putting it back to how it was.